ConfigurationResolver - Fix doubled directory separator

ideaconnect opened the pull request:
It seems that `php-cs-fixer` displays annoying double (or even triple for disks'
root folders) slashes after main path on both Windows and Debian:

1. Windows
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run soap
Loaded config default.
Using cache file ".php_cs.cache".
   1) soap\\src\client.php
   2) soap\\src\test\client.php
```

2. Windows, root D:
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run d:
Loaded config default.
Using cache file ".php_cs.cache".
   1) D:\\\broken.php
   2) D:\\\sub1\broken.php
   3) D:\\\sub1\sub2\broken.php
   4) D:\\\sub3\broken.php
   5) D:\\\sub3\sub4\broken.php
```

3. Windows, root D:, with trailing slash:
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run d:\
Loaded config default.
Using cache file ".php_cs.cache".
   1) D:\\\broken.php
   2) D:\\\sub1\broken.php
   3) D:\\\sub1\sub2\broken.php
   4) D:\\\sub3\broken.php
   5) D:\\\sub3\sub4\broken.php
```

4. Windows, with trailing slash:
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run soap/
Loaded config default.
Using cache file ".php_cs.cache".
   1) soap\\src\client.php
   2) soap\\src\test\client.php
```

5. Windows, absolute path:
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run C:\Users\bartosz\idct\soap
Loaded config default.
Using cache file ".php_cs.cache".
   1) soap\\src\client.php
   2) soap\\src\test\client.php
```

6. Windows, absolute path, with trailing slash:
```dos
C:\Users\bartosz\idct>php-cs-fixer fix --dry-run C:\Users\bartosz\idct\soap\
Loaded config default.
Using cache file ".php_cs.cache".
   1) soap\\src\client.php
   2) soap\\src\test\client.php
```

7. Debian
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run ../fail
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail//broken.php
   2) /srv/home/bpacholek/fail//sub/broken.php
```

8. Debian, with trailing slash
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run ../fail/
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail//broken.php
   2) /srv/home/bpacholek/fail//sub/broken.php
```

9. Debian, absolute path
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail//broken.php
   2) /srv/home/bpacholek/fail//sub/broken.php
```

10. Debian, absolute path, with trailing slash
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail/
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail//broken.php
   2) /srv/home/bpacholek/fail//sub/broken.php
```

It seems that we need not only disable adding of the trailing slash in the
`resolveFinder` of `ConfigurationResolver` but also remove it when added by
`realpath`.

After the change:

__Windows env__

1. Windows
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run ../soap
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

2. Windows, root D:
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run d:
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) D:\broken.php
   2) D:\sub1\broken.php
   3) D:\sub1\sub2\broken.php
   4) D:\sub3\broken.php
   5) D:\sub3\sub4\broken.php
```

3. Windows, root D:, with trailing slash:
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run d:\
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) D:\broken.php
   2) D:\sub1\broken.php
   3) D:\sub1\sub2\broken.php
   4) D:\sub3\broken.php
   5) D:\sub3\sub4\broken.php
```

4. Windows, with trailing slash:
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run ../soap/
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

5. Windows, absolute path:
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run c:\Users\bartosz\idct\soap
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

6. Windows, absolute path, with trailing slash:
```dos
C:\Users\bartosz\idct\fixer>php php-cs-fixer fix --dry-run c:\Users\bartosz\idct\soap\
Loaded config default from "C:\Users\bartosz\idct\fixer\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

7. Debian:
```bash
Checked all files in 0.269 seconds, 10.500 MB memory used
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run ../fail
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

8. Debian, with trailing slash:
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run ../fail/
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

9. Debian, absolute path:
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

10. Debian, absolute path, with trailing slash:
```bash
[bpacholek@dev01 fixer]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail/
Loaded config default from "/srv/home/bpacholek/fixer/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

---

I believe this could be done in a more elegant way for somebody who knows the
code better, yet it solves the problem. _Sorry if I missed something._

`HeaderCommentFixerTest`, `NoWhitespaceInBlankLineFixerTest`, `NoEmptyStatementFixerTest` tests failures are not related to my changes.

I used the branch `2.4` as it is the default one.

ideaconnect commented:
Which case is it? Pointing a direction would be much appreciated - especially that it sounds like you already know these cases.

julienfalque commented on src/Console/ConfigurationResolver.php:
```diff
@@ -737,7 +737,7 @@ function ($path) {
                 $pathsByType['file'][] = $path;
             } else {
-                $pathsByType['dir'][] = $path.DIRECTORY_SEPARATOR;
+                $pathsByType['dir'][] = substr($path, -1, 1) !== DIRECTORY_SEPARATOR ? $path : substr($path, 0, strlen($path) - 1);
```
This will fail if the path is `/`: the value will be an empty string, resulting in an exception being thrown:
```
The "" directory does not exist.
```

julienfalque commented:
Ah sorry, I did write a comment on the changed line mentioning the failing case but it looks like something went wrong when I submitted it and I didn't notice. I just wrote it again.

keradus commented:
sadly, cannot say anything. you claim to fix sth, but there is no test showing what you are fixing :(

julienfalque commented:
I think this part is not covered by any test and it might be hard to write one. I checked the branch out and tested manually.

ideaconnect commented:
Additionally for root dir case `/` it is an issue with `Symfony/Finder` itself:
https://github.com/symfony/finder/blob/master/Iterator/RecursiveDirectoryIterator.php#L79

Causes `getPathname` to return `//` for root references. For absolute paths can be replaced with `getRealPath`

ideaconnect commented:
```php
$finder = new Finder();
$iterator = $finder->files()->in('/');

foreach ($iterator as $file)
{
  print $file->getPathname()."\n";
```

with latest `Finder` gives:

```bash
//client.php
//findertest/test.php
//test2/sub2.php
```

in my test environment.

So until that is solved in Finder a simple workaround should/can be applied in fixer.

ideaconnect commented:
Updated. With `/` there was an additional problem with `getPathname`
returning doubled value, yet this seems to be an issue of `Symfony/Finder` itself as explained above.

Updated the previous set of tests with test of root folder on Ubuntu:

1. Windows
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run ../soap
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

2. Windows, root D:
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run d:
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) D:\broken.php
   2) D:\sub1\broken.php
   3) D:\sub1\sub2\broken.php
   4) D:\sub3\broken.php
   5) D:\sub3\sub4\broken.php
```

3. Windows, root D:, with trailing slash:
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run d:\
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) D:\broken.php
   2) D:\sub1\broken.php
   3) D:\sub1\sub2\broken.php
   4) D:\sub3\broken.php
   5) D:\sub3\sub4\broken.php
```

4. Windows, with trailing slash:
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run ../soap/
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

5. Windows, absolute path:
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run C:\Users\bartosz\idct\soap
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

6. Windows, absolute path, with trailing slash:
```dos
C:\Users\bartosz\idct\fix22>php php-cs-fixer fix --dry-run C:\Users\bartosz\idct\soap\
Loaded config default from "C:\Users\bartosz\idct\fix22\.php_cs.dist".
   1) C:\Users\bartosz\idct\soap\src\client.php
   2) C:\Users\bartosz\idct\soap\src\test\client.php
```

7. Debian
```bash
[bpacholek@dev01 fixer_v22]$ ./php-cs-fixer fix --dry-run ../fail
Loaded config default from "/srv/home/bpacholek/fixer_v22/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

8. Debian, with trailing slash
```bash
[bpacholek@dev01 fixer_v22]$ ./php-cs-fixer fix --dry-run ../fail/
Loaded config default from "/srv/home/bpacholek/fixer_v22/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

9. Debian, absolute path
```bash
[bpacholek@dev01 fixer_v22]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail
Loaded config default from "/srv/home/bpacholek/fixer_v22/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

10. Debian, absolute path, with trailing slash
```bash
[bpacholek@dev01 fixer_v22]$ ./php-cs-fixer fix --dry-run /srv/home/bpacholek/fail/
Loaded config default from "/srv/home/bpacholek/fixer_v22/.php_cs.dist".
   1) /srv/home/bpacholek/fail/broken.php
   2) /srv/home/bpacholek/fail/sub/broken.php
```

11. Ubuntu, root path
```bash
bartosz@DESKTOP-5A99J7Q:/mnt/c/Users/bartosz/idct/fix22$ sudo php php-cs-fixer fix --dry-run /
Loaded config default from "/mnt/c/Users/bartosz/idct/fix22/.php_cs.dist".
   1) /client.php
   2) /findertest/test.php
```

Switched to branch `2.2`
Tried all scenarios without `--dry-run` too.

As Dariusz ( @keradus ) mentioned above this part seems not to be covered by any unit tests.

@julienfalque this could be covered by a unit test which would overwrite `$collection`, return a mock with fake data from `fixFile` and finally comparing results of `fix()` ( `$changed` ) as the test pass or fail.

This is a very minor issue so thank you very much for immediate participation!

keradus commented:
> As Dariusz ( @keradus ) mentioned above this part seems not to be covered by any unit tests.

then add unit test that will fail on 2.2 but pass on your PR.
without it, this change will not get approved.

ideaconnect commented:
Your tone is not very friendly here to be honest. I reported a bug, provided a fix and you are :

doubts in `you claim to fix sth`, ordering tone in `without it, this change will not get approved.`  - the sentence should sound: "Great, thanks for participation, just please finish that up with a unit test as it seems we have missed a part of the code in the current set - sorry for that!"

keradus commented:
Sorry, I'm just tried of replying same things all the time :(

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.4/CONTRIBUTING.md#quick-guide
> If you are adding functionality or fixing a bug - add a test!

ideaconnect commented:
Everybody has a limited set of time: have I anywhere said that I do not want to make the test or update existing ones? No... Just first wanted to confirm whenever this fix is to be accepted to avoid writing pointless code. 

Plus it is very easy to notice this bug in your code: the only thing required is to run the application on any Windows or Debian (+Ubuntu), have not tested on OSX as I do not have access. So I am a bit amazed that you have not noticed it before.

Nonetheless nobody has bad intentions here - thanks @julienfalque for confirmation, I shall finish this PR with tests update. @keradus as you are a member of "FriendsOfPHP" I just kindly suggest to be more "friendly" as everybody who participate here is most likely trying to do something good for your code and really really cause no harm.

keradus commented:
> have not tested on OSX as I do not have access

we do run tests on Ubuntu, Windows and OSX. go ahead ;)

> So I am a bit amazed that you have not noticed it before.

probably it was not an issue for anyone yet, as `foo//bar` is the same path as `foo//bar

> I just kindly suggest

sure, I will try :+1:
in return, I ask to check out guidelines for projects you are contributing to, so you would avoid such a discussion in future ;)

keradus commented on src/Runner/Runner.php:
```diff
@@ -134,8 +134,16 @@ public function fix()
+                $path = $file->getPathname();
+
+                /* workaround for Symfony/Finder setting doubled slash for root folders in
+                SplFileInfo. See `RecursiveDirectoryIterator::current()`. */
```
is it known bug of `Symfony/Finder` ? if so, please link a issue.
if not, please make an issue and link it.

ideaconnect replied:
Have not found officialy reported one, but also not seen all. Made a fork already, planning to make a PR there with a fix - when open will link it. I should have some time tomorrow so hopefully will grab this on.

keradus commented on src/Runner/Runner.php:
```diff
@@ -134,8 +134,16 @@ public function fix()
+
+                /* workaround for Symfony/Finder setting doubled slash for root folders in
+                SplFileInfo. See `RecursiveDirectoryIterator::current()`. */
+                if (substr($path, 0, 2) === '//') {
```
not needed for `\\` ?

ideaconnect replied:
I do not think so as cannot really happen: PHP cannot handle microsoft's special paths starting with `\\` (for network resources) and cannot imagine any other usage of `\\` - if you know any plz let me know. Sorry, cannot bring any out of my head at the moment.

keradus replied:
you provided examples like:
> `soap\\src\client.php`
> ` /srv/home/bpacholek/fail//broken.php`

and then provided code for `//` but not `\\`. probably with tests it will be easier to discuss.

ideaconnect replied:
The example above is fixing the case for double `//` at the start of a path which is caused by finder's bug. Doubled `DIRECTORY_SEPARATOR` bug is caused by `$pathsByType['dir'][] = $path.DIRECTORY_SEPARATOR;` fixed by ` $pathsByType['dir'][] = ($path === '/' || substr($path, -1, 1) !== DIRECTORY_SEPARATOR) ? $path : substr($path, 0, strlen($path) - 1);`

keradus replied:
let's discuss it after adding tests covering all cases

keradus commented:
ping @ideaconnect

SpacePossum commented:
I feel this is close to a bug resolved and therefor would like to this see this finished :)
@ideaconnect I know this PR wasn't received in the friendly matter you would expected it to be, I hope however you still feel like helping us (and the users) out here after the talks here.
Please let us know if you can pick this one up again!

keradus commented:
closing, as:
- after whole discussion we were still walking around of what we think rather than creating new tests
- CI is failing
- looks more like Sf component issue
- lack of activity

if one is willing to pick it up, please create new PR.
